velox/exec/HashBuild.cpp

```diff
@@ -779,11 +778,62 @@ bool HashBuild::finishHashBuild() {
   return true;
 }
 
-void HashBuild::ensureTableFits(uint64_t numRows) {
+void HashBuild::ensureTableFits(
```

Comment:
```
void HashBuild::ensureTableFits(
    const std::vector<HashBuild*>& otherBuilds,
    const std::vector<std::unique_ptr<BaseHashTable>>& otherTables,
    bool isParallelJoin,
    bool allowDuplicateRows) {
  // NOTE: we don't need memory reservation if all the partitions have been
  // spilled as nothing need to be built.
  if (!spillEnabled() || spiller_ == nullptr || spiller_->isAllSpilled()) {
    return;
  }
  VELOX_CHECK_EQ(otherBuilds.size(), otherTables.size());

  // Test-only spill path.
  if (testingTriggerSpill(pool()->name())) {
    Operator::ReclaimableSectionGuard guard(this);
    memory::testingRunArbitration(pool());
    return;
  }   

  const auto dupRowOverheadBytes =
      sizeof(char*) + sizeof(NextRowVector);  

  uint64_t totalNumRows{0};
  uint64_t lastBuildBytesToReserve{0};
  {
    std::lock_guard<std::mutex> l(mutex_);
    const auto numRows = table_->rows()->numRows();
    totalNumRows += numRows;
    if (allowDuplicateRows) {
       lastBuildBytesToReserve += numRows * dupRowOverheadBytes;
    }
  }

  for (auto i = 0; i < otherTables.size(); i++) {
    auto& otherTable = otherTables[i];
    VELOX_CHECK_NOT_NULL(otherTable);
    auto& otherBuild = otherBuilds[i];
    const auto& rowContainer = otherTable->rows();
    int64_t numRows{0};
    {
        lock guard on other build
        numRows = rowContainer->numRows();
    }
    if (numRows == 0) {
      continue;
    }

    totalNumRows += numRows;
    if (!allowDuplicateRows) {
       continue;
    }
    
    const auto dupRowBytesToReserve = numRows * dupRowOverheadBytes;
     if (!isParallelJoin) {
        lastBuildBytesToReserve += dupRowBytesToReserve;
        continue;
     }
     
     Operator::ReclaimableSectionGuard guard(otherBuild);
      auto* otherPool = otherBuild->pool();

      // Reserve memory for memory allocations for next-row-vectors in
      // otherBuild operators if it is parallel join build. Otherwise all
      // next-row-vectors shall be allocated from the last build operator.
      if (!otherPool->maybeReserve(dupRowBytesToReserve)) {
        LOG(WARNING)
            << "Failed to reserve " << succinctBytes(dupRowBytesToReserve)
            << " for for duplicate row memory allocation from non-last memory pool "
            << otherPool->name()
            << ", usage: " << succinctBytes(otherPool->usedBytes())
            << ", reservation: " << succinctBytes(otherPool->reservedBytes());
    }
  }

  if (totalNumRows == 0) {
    return;
  }

  // NOTE: reserve a bit more memory to consider the extra memory used for
  // parallel table build operation.
  lastBuildBytesToReserve += table_->estimateHashTableSize(totalNumRows) * 1.1;
  {
    Operator::ReclaimableSectionGuard guard(this);
    TestValue::adjust(
        "facebook::velox::exec::HashBuild::ensureTableFits", this);
    if (pool()->maybeReserve(lastBuildBytesToReserve)) {
      return;
    }
  }

  LOG(WARNING) << "Failed to reserve " << succinctBytes(lastBuildBytesToReserve)
               << " for last build memory pool " << pool()->name()
               << ", usage: " << succinctBytes(pool()->usedBytes())
               << ", reservation: " << succinctBytes(pool()->reservedBytes());
}
```